Text anchor positions have been added.

nuald opened the pull request:
A follow-up PR regarding text alignment. The main change is a new attribute added to the `TextStyle` - the text anchor position that specifies the position of the anchor point (used in `draw_text`) relative to the text itself. I had to dismiss my original idea to use rectangles as there is a `Text` element that is a `Drawable`, therefore need to draw itself using specified points. As a bonus, the current implementation is backward compatible.

Unfortunately, completely get rid of the text size estimation isn't possible yet, therefore there are the use-cases there the text won't be properly aligned yet:
- title as it uses the text height to adjust the drawing area;
- series labels as they need to calculate the bound rectangle;
- multiline text. The current implementation is not quite compatible with the client side rendering as rather the backend should calculate its layout, not the element itself (for example, SVG would need to utilize `<tspan>` tag for that).

However, it doesn't affect the actual visible results so much (unless series labels border is visible) as all of those elements utilize the margins.

Additional notes:
- Now backends do all text transformations, no fonts themselves (as a side effect, the font `draw` code is much simpler now as no offsets calculations are needed). It affect multiline element rendering if rotated, however, it doesn't properly rotate even in the original code;
- `should_draw` check is gone from axis drawing as we can't reliable determine the text bounds, so we utilize clipping for that now (verified and fixed for all backends). Also I guess that users prefer partially visible text than completely hidden. However, it would be nice to have support for mesh margins, so we could put some extra spacing thus avoiding clipping;
- the rendering results changed for all examples, so I guess they need to be updated;
- SVG for MS IE/Current Edge has issues because they don't support all tags from the spec. However, as Edge is switching to Chromium rendering engine, it won't be an issue in near future (I've verified with beta Edge). If it could be an issue, I can try to find some workarounds;
- fixed `draw_pixel` in `bitmap` backend to normalize alpha as the `animation` example send overflowed float values to it.
- `console` example have a gap now between bottom labels and the axis, but it's not because of the miscalculations, but rather because ticks are not drawn there (the `draw_line` function doesn't draw vertical lines with 1 `pixel` width).

38 commented on src/drawing/backend_impl/bitmap.rs:
```diff
@@ -231,6 +231,11 @@ pub trait PixelFormat: Sized {
                         *buf.get_unchecked_mut(base + idx) = Self::byte_at(r, g, b, 0, idx);
                     }
                 } else {
+                    let alpha = alpha.min(1.0).max(0.0);
```
Just wondering isn't this already checked to be less than 1.0 already? There might be a overflow, but it seems what we should do is changing the if condition from `1.0` to `255.0 / 256.0`?

38 commented on examples/console.rs:
```diff
@@ -127,10 +128,22 @@ impl DrawingBackend for TextDrawingBackend {
+        let dy = match style.pos.v_pos {
+            VPos::Top => 0,
+            VPos::Center => -height / 2,
+            VPos::Bottom => -height,
```
I went through the code, it seems the new code doesn't really guarantee the text isn't in the bouncing box.  Since the decedent part will be always out of box, and this should be avoid.

38 commented on src/chart/context.rs:
```diff
@@ -549,81 +516,70 @@ impl<'a, DB: DrawingBackend, X: Ranged, Y: Ranged> ChartContext<'a, DB, RangedCo
+                (cx, cy + label_offset)
             };
 
-            if should_draw {
```
And one reason for this `should_draw`check is we need to avoid the tag is drawn out of it's own area, but if we only relies on the backend, a super long label may overlap with other drawing area on the right side (for example). And I think we should avoid this

nuald replied:
It could be confusing to users - if super long label is not drawn at all, they may think it's a bug in their code. I think the proper fix would be implement clipping in the drawing area, but I'm not sure how to properly do it with the current architecture (as drawing backend is not supposed to know about the drawing area). I guess it's possible to add a new API call for the backend like `set_clipping_region`, but it's rather high-level change that I won't risk to add without your explicit approval.

38 replied:
Ok, I think your comment make sense. I think I am Ok with this . But I don't think we don't want the backends aware of any high-level details. Since I feel we should keep high-level and low-level API separate. `set_clipping_region` solution might be some solution, but  I think it needs some amount of work to do and I don't think it's worth us to dig into this.

38 commented on src/style/font/ttf.rs:
```diff
@@ -159,31 +159,25 @@ impl FontData for FontDataInternal {
-
         for g in font.layout(text, scale, point(0.0, 0.0)) {
             if let Some(rect) = g.pixel_bounding_box() {
-                let x0 = rect.min.x;
```
And I believe this is one of the reason why the descender is out of the box now.

nuald replied:
Yes, it's a regression bug I've introduced, the up-to-dated change has the proper calculations.

nuald commented on examples/boxplot.rs:
```diff
@@ -94,13 +93,14 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             }))?
             .label(label)
-            .legend(move |(x, y)| Rectangle::new([(x - 5, y - 5), (x + 5, y + 5)], style.filled()));
+            .legend(move |(x, y)| Rectangle::new([(x, y - 3), (x + 12, y + 9)], style.filled()));
```
Legend is drawn correctly, I've just changed the rectangle location to be aligned more with the text (original change had a misalignment already, but I hoped that it could be fixed with the proper text rendering, but as I mentioned above, the legends rendering is not easy to fix and out of scope of this PR).

nuald commented:
Few additional notes:
- Is there any specific reason for not using `draw_pixel` for client-side backends (SVG and Canvas)? I've verified, just removing `draw_text` from those backends makes the resulting images are properly rendered with the server-side font. Surely, it has performance penatly and pixels are visible if the image is scaled, but if users require better text positioning, it would be nice to have an option to use the server-side rendering.
- You may notice that cairo has issues with `draw_text` on Windows (`_win32_scaled_font_set_world_transform` error with _helpful_ description "error occurred in the Windows Graphics Device Interface" that comes from WinAPI itself), so with the new change full test suite is going to fail quite often there (as a lot of `draw_text` calls are used). I've dig it a little, but couldn't find an easy fix:
  - it's not because tests are multithreaded (I've tested a global mutex for all font/text related operations);
  - setting fonts explicitly instead of auto-selecting doesn't help either (and it could be an actual bug in the cairo-win32-font as it doesn't directly use the font from the context, but looks like it creates additional ones and have it messed up sometimes);
  - the proper fix could be avoiding _toy_ functions and using proper FreeType fonts loaded by font-kit (in that case cairo-win32-font won't be probably even used), but it requires some additional work which I'm not sure anybody would care about :)

38 commented:
>     * Is there any specific reason for not using `draw_pixel` for client-side backends (SVG and Canvas)?

For this one, I think it's an important feature to use the client-side rendered text. Otherwise, some of the client is very slow and we just lose the reason for using SVG image format at all. But yes, we actually can render a bitmap image and blit it onto the SVG image but I think this is another story. 

> You may notice that cairo has issues with draw_text on Windows 
I am not sure, since I don't really have a Windows machine on my hand. But I am not sure if there is anybody who really care about the test case fails on Windows. It's not bothering me at all, but I would may want to make sure that it's something really related to cairo-rs bug specifically on Window. If this is the case I  think I can make it a low priority task.

38 commented:
![image](https://user-images.githubusercontent.com/3023614/69503884-829a5300-0edb-11ea-9c69-d9e017efbc4a.png)
Another regression I realized is the tick labels on the right side, seems not properly aligned.

38 commented on src/drawing/backend_impl/bitmap.rs:
```diff
@@ -231,7 +231,7 @@ pub trait PixelFormat: Sized {
                     }
                 } else {
-                    let alpha = alpha.min(1.0).max(0.0);
+                    let alpha = alpha.min(255.0 / 256.0).max(0.0);
```
I believe we can juse use if alpha >= 255.0 / 256.0 as the check for if we need to directly override the pixel.

nuald replied:
Could you please elaborate? The then-branch above has the check for overriding the pixel (alpha >= 1.0), do I miss something here?

38 replied:
I think the reason why it can overflow is because `(alpha * 256.0).floor()` returns a 256.0, this might be possible since we have floating point issues. But if you dig into the logic at this point, we actually split the range (0.0..1.0] to 256 different alpha levels, and which means anything is greater than 255.0 / 256.0 can be treated as directly overriding, right? 
By doing so, I believe there won't be any case causing overflow anymore.

Please let me know if I was wrong.

nuald replied:
Sorry, looks like I was not clear in the original PR message - the overflow is not happening here, but rather in the `animation` example. The fix is required to properly address invalid `alpha` values. I've added `test_draw_pixel_alphas` tests which you can use to verify the behaviour - without the fix, the bitmap code is crashing.

38 commented on src/chart/context.rs:
```diff
@@ -549,57 +516,54 @@ impl<'a, DB: DrawingBackend, X: Ranged, Y: Ranged> ChartContext<'a, DB, RangedCo
                     // Right
                     (dx, dy) if dx > 0 && dy == 0 => {
-                        (right_alignment + label_dist - w as i32, *p - y0)
+                        (label_dist, *p - y0, HPos::Left, VPos::Center)
```
Could we use right alignment here?

nuald replied:
The labels on the right side of the chart use the left alignment as before, nothing changed here. The name `right_alignment` was quite confusing as it was used in a meaning "the max width of the labels" (therefore the calculations produced the proper offset for all labels being on the same vertical line with the distance of `label_dist` from the axis).

38 replied:
I believe there's a change here. Since before the change, we just make sure the right aligned for all the labels. 
Sorry about the confusing, but this is what I mean.

Before the change:
![image](https://user-images.githubusercontent.com/3023614/69563115-b625bd80-0f6d-11ea-96db-0c2ca0948483.png)

After the change:
![image](https://user-images.githubusercontent.com/3023614/69563160-ca69ba80-0f6d-11ea-92d0-efffc78cb60f.png)

I still think getting the max width is some step we should do here

nuald replied:
Oh, I see, actually I was thinking it's a bug and never realized that it's intended :) Please let me explain - usually on the charts the right labels are aligned by the left side as the width could differ a lot. For your use-case the right alignment looks better because the digits are aligned (and for usual linear scales that probably would be better). However, for logarithmic and category labels it could be not so nice (for example, legitimate IPv4/IPv6 addresses):

![boxplot](https://user-images.githubusercontent.com/382655/69571606-a7e09d00-0f7f-11ea-9f3a-1ea4a87ab993.png)

The mapping between the ticks and labels is not clearly visible. I could suggest 2 options:
1. Add an option for the alignment (now sure how the API would look like though);
2. Align based on the label source (category vs numeric etc, but again not sure is it even possible).

Unfortunately, I need the left alignment (as labels on my boxplot chart have quite different widths) and would prefer to avoid any hacks and find the proper solution.

38 commented on src/drawing/backend_impl/svg.rs:
```diff
@@ -247,24 +248,24 @@ impl<'a> DrawingBackend for SVGBackend<'a> {
+            HPos::Center => "middle",
         };
+        let dominant_baseline = match style.pos.v_pos {
+            VPos::Top => "hanging",
```
Just tried with `eog`, which is `librsvg` based SVG rendering, it seems it doesn't support this. So could we think about not supporting the vertical anchor point for now, because I think we should generate SVG that renders properly in most of the clients.

38 replied:
Or at least, we don't rely on `dominant_baseline` for a while until this is widely supported.

nuald replied:
Thanks, I'll add the tests and try to find a solution.

38 commented on src/drawing/backend.rs:
```diff
@@ -191,15 +192,37 @@ pub trait DrawingBackend: Sized {
+        };
+        let dy = match style.pos.v_pos {
+            VPos::Top => height,
+            VPos::Center => height / 2,
```
I believe you have changed the `Font::draw` to use the baseline as coordinate, this is perfect. But if this is the case, we may just have a very small offset when we use the middle anchor point. I suspect this is related to the small offset in all the Y-axis label
![image](https://user-images.githubusercontent.com/3023614/69504551-0b67bd80-0ee1-11ea-89b9-58e45852004d.png)

38 commented on src/style/font/font_desc.rs:
```diff
@@ -255,7 +256,7 @@ impl<'a> FontDesc<'a> {
             font: self.clone(),
             color: color.to_rgba(),
-            alignment: TextAlignment::Left,
+            pos: Pos::default(),
```
Not about this one, but I believe after the change, we can get rid of `FontTransform::offset` right? Because we have better model that rotating the text based on the anchor point as the center.

nuald replied:
Yes, I'll remove it.

nuald commented:
You're absolutely right that there are some issues with the middle anchor points, but it's not a regression bug, it's something unexpected I couldn't see before until you pointed it to me (but maybe you know the reason). I've modified the tests to draw the original layout_box returned by the `rusttype` font itself (as red circles for min and max `pixel_bounding_box`):

![layout_box](https://user-images.githubusercontent.com/382655/69511411-36f8a100-0efd-11ea-8291-293cd92e1953.png)

As you could see the `x` bounds are correct, but `y` are not. I'm going to research it, and I hope it's not rusttype bug.

nuald commented:
I've re-verified `rusttype` bbox from the scratch, and while it does have minor differences from the actual pixels it draws, the problem was in my y-values calculations, I've fixed it.

SVG has been fixed too to use `ex`-based lengths with `dy` instead of `dominant_baseline` (and actually, it's improved the accuracy a little). Unfortunately, it affects titles even more now, but I doubt we can do anything with it without changing/adding some API.

38 commented on src/drawing/backend_impl/svg.rs:
```diff
@@ -247,24 +248,36 @@ impl<'a> DrawingBackend for SVGBackend<'a> {
+        let text_height = text_max_y as f32 - text_min_y as f32;
+        let text_middle = text_min_y as f32 + text_height / 2.0;
+        let scaled_height = text_height / x_height;
+        let offset = (text_middle - x_middle) / x_height;
```
I think the offset computing has come problem. My understanding is this offset actually makes the real bottom of the bouncing box at the `pos` location, which is exactly other parts of the code defines the locations. 
If this is the case, shouldn't it be 
```rust
offset = (text_max_y - x_max_y) / x_height;
```
As long as "x" doesn't have any part under the baseline, otherwise we may ending up with half of the size between baseline and descender. 

Correct me if I was wrong.

38 replied:
Also, I am ok with adding an interface that returns the relative descender size if you think this makes things easier. Since I believe rusttype have some way to returns the metrics of the font.

nuald replied:
You're right that the offset calculations are slightly incorrect, but it's not as simple as just using `max_y` as it makes the vertical center incorrect with the current code. I'll update the calculations and tests with the different ascends/descends. And I doubt that any interface changes are required, `layout_box` is working good enough, it's just a matter of proper testing to finalize the calculations.

38 replied:
Yes, dig into this a little bit, I believe "dy=1em" is same as "dominant-baseline=hanging"
"dy=0.5ex" is same as "dominat-baseline=middle"  and
"dy=-0.5ex" is almost same as putting the dominant baseline to the actual bottom of the text.

How do you think about this. 

Also, I would like to say rusttype underestimates the text size a lot, so I think we should adjust the SVG's font size.

38 replied:
To explain the magic numbera 0.75, the difference between rusttype scaling and SVG scaling is SVG uses em in pixels as font size, but rusttype uses the total height of text as the size. 
So in SVG sizing model, descender isn't a part of font size, but in rusttype it also counted as the font size. 

That gives us 1em = 1 pixel is same as rusttype size 1.3. (Assuming we have descender height of 30%. And I believe the typical descender height is some value around 20% to 30%). 
That give us the factor 1 / 1.3 = 0.76 which is almost the magic number I come up with.

38 replied:
If this is the case, we actually have a pretty reliable way to define the anchor point. But the problem we have to facing is the font positioning, and my guess is this is impossible with solely SVG. 

For example, for the title spacing, we have to know the font size to decide the size of drawing area. The text size is something we have to make an reasonable guess to make it work. 

So our remaining problem could be rusttype isn't follow the SVG sizing model and that's the root for the remaining offset in the font layout. 

I actually tried render text using font-kit instead of rusttype (which I can remove a few dependency as well), it seems font-kit renders text in exact same way a browser does. Thus once we have make the change to replace rusttype, I believe we could have pretty stable SVG text layout then. 

The only thing of this plan is, changing the sizing model is definitely a breaking change which we have to wait until the next major release. 

But I believe other part of the PR seems good to me, except the label text on the right label area. 
(And I don't think this should be considered as breaking change and I believe it's ready to merge).

Sorry about the messy font handling code, as you may realized, I am just learning typography while I was creating this crate and it's a really good lesson I have learned.  

Please let me know you have any thought, and thank you so much for clean this up!

38 replied:
And update, according to wikipedia, typically, the descender height is 46% of em,which give us `dy=0.76em`
This gives us pretty much accurate result, I believe in most of the case
https://jsfiddle.net/rjmtz602/

nuald replied:
Thank you for the research and numbers, I've updated the code and indeed it looks very good.

nuald commented:
I've updated the tests to verify drawing the text with the different ascends/descends. SVG and Canvas don't have the perfect pixel alignments, but nevertheless look good enough from my point of view.

Also I've discovered and fixed some minor regression issues with the category coord.

nuald commented:
If you don't mind, I'd prefer that you would polish the remaining items - you obviously know the context better, plus you may align your changes with your project vision. Also, I don't have the Linux desktop, only server, and actually use WSL with XServer compiled for Windows, and it has some minor issues with displaying the content (as I have the high DPI monitor that applies some scaling). So mostly all the changes are verified on Windows only (and partially on macOS), and I could definitely miss some use-cases.

38 commented:
Of course, I verified this on Linux and MacOS. I believe the SVG rendering is better than before after the change.

This is a quite big improvement and I really appreciate for your help! 

Cheers!

38 commented:
Finally get this merged, nice work! And I am going to ship a minor release in a week or two hopefully. Thanks for making Plotters better!
